cmd/e2e-test/carbon_clickhouse.go

```diff
@@ -0,0 +1,124 @@
+package main
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"path"
+	"path/filepath"
+	"strconv"
+	"syscall"
+	"text/template"
+
+	"github.com/phayes/freeport"
+)
+
+type CarbonClickhouse struct {
+	Binary    string `toml:"binary"`
+	ConfigTpl string `toml:"template"`
+
+	storeDir   string    `toml:"-"`
+	configFile string    `toml:"-"`
+	address    string    `toml:"-"`
+	cmd        *exec.Cmd `toml:"-"`
+}
+
+func (c *CarbonClickhouse) Start(clickhouseAddr string) error {
+	if c.cmd != nil {
+		return fmt.Errorf("carbon-clickhouse already started")
+	}
+
+	if len(c.Binary) == 0 {
+		c.Binary = "./carbon-clickhouse"
+	}
+	if len(c.ConfigTpl) == 0 {
+		return fmt.Errorf("carbon-clickhouse config template not set")
+	}
+
+	var err error
+
+	c.storeDir, err = ioutil.TempDir("", "carbon-clickhouse")
+	if err != nil {
+		return err
+	}
+	port, err := freeport.GetFreePort()
+	if err != nil {
+		c.Cleanup()
```

Comment:
Here what I would suggest for reducing the code:

- Use `func (c *CarbonClickhouse) Start(clickhouseAddr string) (err error) {` as a func declaration
- Add a little function, that will Clean, if `err != nil`. Here the playground that shows, how it works https://play.golang.org/p/VyMjpii6HTw